File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.py
class Solution:
    def reverseWords(self, s: str) -> str:
        li = []
        if s[0].isalnum():
                li.append(0)
        for i in range(1,len(s)):
            if s[i-1]==' ' and s[i].isalnum():
                li.append(i)
            if s[i-1].isalnum() and s[i] == ' ':
                li.append(i-1)
        if s[len(s)-1].isalnum():
            li.append(len(s)-1)
        newli = []
        for i in range(len(li)-1,0,-2):
            start = li[i-1]
            end = li[i]
            word = s[start:end+1]
            newli.append(word)  # reverse each word
        return " ".join(newli)

File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.java
class Solution {
    public String reverseWords(String s) {
        StringBuilder sb = new StringBuilder();
        int i = s.length()-1 ;
        while(i>=0){
            while(i>=0 && s.charAt(i)==' ')i--;
            int start = i;
            while(i>=0 && s.charAt(i)!=' ')i--;
            sb.append(s.substring(i+1,start+1));
            sb.append(" ");
        }
        return sb.toString().trim();
    }
}